### DISCUSSION TITLE
afl-compiler-rt.o.c: CMPLOG relative function return address problem

### DISCUSSION THREAD

**Discussion by che30122** [AFLplusplus/AFLplusplus#1402] [Q&A] [answered]
See the code snippet below.

uintptr_t k = (uintptr_t)__builtin_return_address(0);
  k = (uintptr_t)(default_hash((u8 *)&k, sizeof(uintptr_t)) & (CMP_MAP_W - 1));

I thought that if the program under test with PIE enable, that fuzzer will get different hash value, even the hooked instruction in program under test is same.
Is that correct?

**Comment by vanhauser-thc** [comment, marked as answer]
yes/no.

1) afl-fuzz will complain on startup if randomize_va_space is not set to 0. this would disable random addresses.
2) it does not matter because a fork server is used, so for every fuzz attempt the addresses are stable as the libraries are already loaded so they dont change while running.

**Reply by che30122** [reply]
Thanks a lot.

A futher question is that I saw the code snippet below.

 afl->cmplog_fsrv.trace_bits = afl->fsrv.trace_bits;

So even if I run cmplog_fsrv, it still can update coverage feedback.

Is that correct?
